### PR TITLE
CMake: Test link interface again

### LINKED ISSUE #15077: vanilla Ubuntu 20.04 linker errors with OpenMPI
I am seeing linker errors 
```
[ 85%] Building CXX object examples/CMakeFiles/example_step_29_release.dir/step-29/step-29.cc.o
/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: error: undefined reference to 'opal_uses_threads'
/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: error: undefined reference to 'opal_class_initialize'
/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: error: undefined reference to 'opal_list_item_t_class'
/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so: error: undefined reference to 'opal_class_init_epoch'
collect2: error: ld returned 1 exit status
```
with the current deal.II master. This is because ``-fuse-ld=gold`` is picked up even though it can't correctly link. We used to drop gold because of the "have usable flags check", which doesn't work anymore?

### PR DESCRIPTION
**Pull request by masterleinad**
Part of https://github.com/dealii/dealii/issues/12374. Fixes #15077.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +120,8 @@ foreach(_file ${_configure_files})
   verbose_include(${_file})
 endforeach()
 
+verbose_include(${CMAKE_SOURCE_DIR}/cmake/setup_finalize.cmake)
```
Comment (tamiko) [inline CMakeLists.txt:123]: And it's back at its rightful place :-D

File: cmake/setup_finalize.cmake [hunk 2]
```diff
@@ -55,12 +55,64 @@ endforeach()
...
+    foreach(_flags
+        DEAL_II_LINKER_FLAGS DEAL_II_LINKER_FLAGS_${build}
+        BASE_LINKER_FLAGS BASE_LINKER_FLAGS_${build}
+        )
+      string(REPLACE "${_linker_flag}" "${_replacement_flag}"
+        ${_flags} "${${_flags}}"
+        )
+    endforeach()
+    set(${_variable} FALSE CACHE INTERNAL "" FORCE)
+    set(${_variable} FALSE)
```
Comment (tamiko) [inline cmake/setup_finalize.cmake:81]: Would you mind refactoring the `set(${_variable ...)` part into its own macro?
It's a bit odd to put this into the logic above.

File: cmake/setup_finalize.cmake [hunk 3]
```diff
@@ -55,12 +55,68 @@ endforeach()
...
+      string(REPLACE "${_linker_flag}" "${_replacement_flag}"
+        ${_flags} "${${_flags}}"
+        )
+    endforeach()
+    _set_cache_variable(_variable FALSE)
+  endmacro()
+
+  _check_linker_flags()
+
+  if(NOT DEAL_II_HAVE_USABLE_FLAGS_${build} AND DEAL_II_COMPILER_HAS_FUSE_LD_LLD)
```
Comment (tjhei) [inline cmake/setup_finalize.cmake:90]: what about mold?

**Review by tjhei — APPROVED** [review]
Awesome, thank you. Should fix #15077.

File: cmake/setup_finalize.cmake (hunk 3 above)
Comment (masterleinad) [reply cmake/setup_finalize.cmake:104]: I added another check for it now.

Comment (tjhei) [reply cmake/setup_finalize.cmake:104]: perfect.

**Review by tjhei — APPROVED** [review]

**Review by tamiko — APPROVED** [review]

**Pull request merged**